source/opt/unify_const_pass.cpp

```diff
@@ -0,0 +1,184 @@
+  analysis::DefUseManager def_use_mgr(module);
+
+  for (ir::Instruction& inst : module->types_values()) {
+    // Do not handle the instruction when there are decorations upon the result
+    // id.
+    if (def_use_mgr.GetAnnotations(inst.result_id()).size() != 0) {
+      continue;
+    }
+
+    // The overall algorithm is to store the result ids of all the eligible
+    // constants encountered so far in a trie. For a constant defining
+    // instruction under consideration, use its opcode, result type id and
+    // words in operands as an array of keys to lookup the trie. If a result id
+    // can be found for that array of keys, a constant with exactly the same
+    // value must has been defined before, the constant under processing
+    // should be replaced by the constant previously defined. If no such result
+    // id can be found for that array of keys, this must be the first time a
+    // constant with its value be defined, we then create a new trie node to
+    // store the result id with the keys. When replacing a duplicated constant
+    // with a previously defined constant, all the uses of the duplicated
+    // constant, which must be placed after the duplicated constant defining
+    // instruction, will be updated. This way, the descendants of the
+    // previously defined constant and the duplicated constant will both refer
+    // to the previously defined constant. So that the operand ids which are
+    // used in key arrays will be the ids of the unified constants, when
+    // processing is up to a descendant. This makes comparing the key array
+    // always valid for judging duplication.
+    switch (inst.opcode()) {
+      case SpvOp::SpvOpConstantTrue:
+      case SpvOp::SpvOpConstantFalse:
+      case SpvOp::SpvOpConstant:
+      case SpvOp::SpvOpConstantNull:
+      case SpvOp::SpvOpConstantSampler:
+      case SpvOp::SpvOpConstantComposite:
+      // Only spec constants defined with OpSpecConstantOp and
+      // OpSpecConstantComposite should be processed in this pass. Spec
+      // constants defined with OpSpecConstant{|True|False} are decorated with
+      // 'SpecId' decoration and all of them should be treated as unique.
+      // 'SpecId' is not applicable to SpecConstants defined with
+      // OpSpecConstant{Op|Composite}, their values are not necessary to be
+      // unique. When all the operands/compoents are the same between two
+      // OpSpecConstant{Op|Composite} results, their result values must be the
+      // same so are unifiable.
+      case SpvOp::SpvOpSpecConstantOp:
+      case SpvOp::SpvOpSpecConstantComposite: {
+        uint32_t id = defined_constants.LookupEquivalentResultFor(inst);
+        if (id != inst.result_id()) {
+          // The constant is a duplicated one, use the cached constant to
+          // replace the uses of this duplicated one, then turn it to nop.
+          def_use_mgr.ReplaceAllUsesWith(inst.result_id(), id);
+          def_use_mgr.KillInst(&inst);
```

Comment:
Yes, agreed.

For newly added instructions, the existing line debug info do not apply at all. If the line-debug instruction's host instruction is killed, a meaningful approach is that still let the line-debug instruction cover whatever is following the original host instruction. Actually that's exactly what we are doing now: #391. (Thanks for fixing the `ForEachInst` and `ToNop` stuff! It's great to see this works out of the box at the moment.)

So I would suggest we just keep what's there right now.